### PR TITLE
CRIU supports Java debugger during checkpoint and restore

### PR DESCRIPTION
**Pull request by JasonFengJ9**
CRIU supports Java debugger during checkpoint and restore

Not suspend `JDWP` threads created in single-threaded mode, handle the `JVMTI` events during `CRIU` checkpoint and restore;
Changed the order of `releaseSafeOrExcusiveVMAccess()`;
Fixed the order of restore phase hooks;
Identified `JDWP` threads via the thread object references instead of checking against the thread name;
Added more tracepoints.
Enabled an OpenJDK `test/jdk/com/sun//jdi/JdwpAttachTest.java`.
* https://github.com/adoptium/aqa-tests/pull/5269

Required by 
* https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/591

Superseded
* https://github.com/eclipse-openj9/openj9/pull/17252

Signed-off-by: Jason Feng <fengj@ca.ibm.com>

### CODE REVIEW

File: runtime/oti/vm_api.h [hunk 1]
```diff
@@ -4380,10 +4380,11 @@ fatalRecursiveStackOverflow(J9VMThread *currentThread);
 * @param privateFlags
 * @param memorySpace
 * @param threadObject the object associated with this thread, or NULL
+* @param threadName string representing thread name, or NULL
```
Comment (keithc-ca) [inline runtime/oti/vm_api.h:4383]: What is this (`allocateVMThread` has no such parameter)?

Comment (JasonFengJ9) [reply runtime/oti/vm_api.h]: Removed the parameter in question.

cherry-picked 
* https://github.com/eclipse-openj9/openj9/pull/19426
Also updated `toggleSuspendOnJavaThreads()` usage & function comments.

**Comment by JasonFengJ9** [comment]
Rebased with the latest change
* https://github.com/eclipse-openj9/openj9/pull/19426

**Comment by JasonFengJ9** [comment]
Resolved the merge conflict.

File: runtime/vm/CRIUHelpers.cpp [hunk 2]
```diff
@@ -1112,57 +1106,64 @@ getNativeString(J9VMThread *currentThread, j9object_t javaString, char **nativeS
...
  *
- * @param[in] currentThread the thread being checked
- * @param[in] toggleDebugThreads if only JWDP debug threads should be toggled
+ * Current usages:
+ * Step 1: suspend - TRUE, toggleDelayedHaltThreads - FALSE
+ *     suspend only those threads NOT marked with the delayed halt flag
+ * Step 2: suspend - TRUE, toggleDelayedHaltThreads - TRUE
+ *     suspend only those threads marked with the delayed halt flag
+ *     previously suspended other threads stay the same
+ * Step 3: suspend - FALSE, toggleDelayedHaltThreads - TRUE
```
[the remark this answers is not in the document]
Comment (JasonFengJ9) [reply runtime/vm/CRIUHelpers.cpp]: Updated the description.

File: runtime/jvmti/jvmtiExtensionMechanism.c [hunk 3]
```diff
@@ -4033,3 +4099,71 @@ jvmtiGetCarrierThread(jvmtiEnv* jvmti_env, ...)
...
+	j9object_t threadObject = NULL;
+	UDATA i = 0;
+	va_list args;
+
+	va_start(args, jvmti_env);
+	thread = va_arg(args, jthread);
+	threadObject = J9_JNI_UNWRAP_REFERENCE(thread);
+	va_end(args);
+
+	for (i = 0; i < vm->checkpointState.javaDebugThreadCount; i++) {
```
[the remark this answers is not in the document]
Comment (JasonFengJ9) [reply runtime/jvmti/jvmtiExtensionMechanism.c:4157]: The L4162-4165 moved the entries after the thread removed forward to fill the hole such that the first `javaDebugThreadCount` entries are not `NULL` which is assumed by `jvmtiAddDebugThreadToCheckpointState` when adding a new thread object.
Also updated the function descriptions.

[the remark this answers is not in the document]
Comment (tajila) [reply runtime/jvmti/jvmtiExtensionMechanism.c:4157]: I understand what its doing. I was saying there is an alternative approach that doesnt involve moving the elements.

[the remark this answers is not in the document]
Comment (tajila) [reply runtime/jvmti/jvmtiExtensionMechanism.c:4157]: Depending on usage patterns thats not necessarily a bad thing. Given the list is a finite size of 10 and I'm going to assume its only being access in single threaded fashion, then it probably doesn't matter either way.

**Comment by tajila** [comment]
jenkins test sanity plinux jdk17

**Comment by pshipton** [comment]
All plinux machines are offline for the foreseeable future.

**Comment by tajila** [comment]
jenkins test sanity alinux64 jdk17

**Comment by tajila** [comment]
@keithc-ca do you have any further comments?

**Comment by keithc-ca** [comment]
I'll have another, closer look.

**Review by keithc-ca — CHANGES_REQUESTED** [review]

File: runtime/include/ibmjvmti.h [hunk 4]
```diff
@@ -106,6 +106,12 @@
 #define COM_SUN_HOTSPOT_EVENTS_VIRTUAL_THREAD_MOUNT "com.sun.hotspot.events.VirtualThreadMount"
 #define COM_SUN_HOTSPOT_EVENTS_VIRTUAL_THREAD_UNMOUNT "com.sun.hotspot.events.VirtualThreadUnmount"
 
+#define OPENJ9_EVENT_VM_CHECKPOINT "openj9.criu.events.VMCheckpoint"
+#define OPENJ9_EVENT_VM_RESTORE "openj9.criu.events.VMRestore"
+
+#define OPENJ9_FUNCTION_ADD_DEBUG_THREAD "openj9.criu.functions.AddDebugThreadToCheckpointState"
+#define OPENJ9_FUNCTION_REMOVE_DEBUG_THREAD "openj9.criu.functions.RemoveDebugThreadFromCheckpointState"
```
[the remark this answers is not in the document]
Comment (JasonFengJ9) [reply runtime/include/ibmjvmti.h]: This does apply to `CRaC` which depends on `CRIU`.
Removed `.criu`, and match the macro identifier roughly.

File: runtime/jvmti/jvmtiExtensionMechanism.c [hunk 5]
```diff
@@ -377,6 +382,27 @@ static const jvmtiParamInfo jvmtiGetCarrierThread_params[] = {
 };
 #endif /* JAVA_SPEC_VERSION >= 19 */
 
+#if defined(J9VM_OPT_CRIU_SUPPORT)
+/* (jvmtiEnv *jvmti_env) */
```
[the remark this answers is not in the document]
Comment (JasonFengJ9) [reply runtime/jvmti/jvmtiExtensionMechanism.c]: Added the second parameter comment, and the blank lines.

File: runtime/vm/CRIUHelpers.cpp [hunk 6]
```diff
@@ -1112,57 +1106,64 @@ getNativeString(J9VMThread *currentThread, j9object_t javaString, char **nativeS
...
  * Caller must first acquire exclusive safepoint or exclusive VMAccess.
  *
  * @param[in] suspend set to true to suspend threads, and false to resume them
- * @param[in] toggleDebugThreads only useful if JDWP is enabled, determines whether to toggle
- * the debug threads or all other threads
+ * @param[in] toggleDelayedHaltThreads whether to toggle threads marked with the delay halt flag or the threads without that flag
  */
 static void
-toggleSuspendOnJavaThreads(J9VMThread *currentThread, BOOLEAN suspend, BOOLEAN toggleDebugThreads)
+toggleSuspendOnJavaThreads(J9VMThread *currentThread, BOOLEAN suspend, BOOLEAN toggleDelayedHaltThreads)
```
[the remark this answers is not in the document]
Comment (JasonFengJ9) [reply runtime/vm/CRIUHelpers.cpp]: Changed it to `bool`.

File: runtime/vm/CRIUHelpers.cpp [hunk 7]
```diff
@@ -1112,57 +1106,64 @@ getNativeString(J9VMThread *currentThread, j9object_t javaString, char **nativeS
...
+	}
 	J9VMThread *walkThread = J9_LINKED_LIST_START_DO(vm->mainThread);
 	while (NULL != walkThread) {
 		if (VM_VMHelpers::threadCanRunJavaCode(walkThread)
-		&& (currentThread != walkThread)
+			&& (currentThread != walkThread)
 		) {
-			if (shouldToggleJavaThread(walkThread, toggleDebugThreads)) {
+			bool isDelayHaltFlagSet = J9_ARE_ANY_BITS_SET(walkThread->privateFlags2, J9_PRIVATE_FLAGS2_DELAY_HALT_FOR_CHECKPOINT);
+			if ((toggleDelayedHaltThreads && isDelayHaltFlagSet) || (!toggleDelayedHaltThreads && !isDelayHaltFlagSet)) {
```
[the remark this answers is not in the document]
Comment (JasonFengJ9) [reply runtime/vm/CRIUHelpers.cpp]: Updated

File: runtime/vm/CRIUHelpers.cpp [hunk 8]
```diff
@@ -1112,57 +1106,65 @@ getNativeString(J9VMThread *currentThread, j9object_t javaString, char **nativeS
...
+	}
 	J9VMThread *walkThread = J9_LINKED_LIST_START_DO(vm->mainThread);
 	while (NULL != walkThread) {
 		if (VM_VMHelpers::threadCanRunJavaCode(walkThread)
-		&& (currentThread != walkThread)
+			&& (currentThread != walkThread)
 		) {
-			if (shouldToggleJavaThread(walkThread, toggleDebugThreads)) {
+			bool isDelayHaltFlagSet = J9_ARE_ANY_BITS_SET(walkThread->privateFlags2, J9_PRIVATE_FLAGS2_DELAY_HALT_FOR_CHECKPOINT);
+			/* suspend/resume when (toggleDelayedHaltThreads && isDelayHaltFlagSet) or (!toggleDelayedHaltThreads && !isDelayHaltFlagSet) */
```
Comment (keithc-ca) [inline runtime/vm/CRIUHelpers.cpp:1152]: I don't think this comment helps much: it's just as much work to understand it as the code that was here before. We could say "toggleDelayedHaltThreads and isDelayHaltFlagSet are both true or both false", but that isn't much better.

Perhaps replacing the pair of `bool` parameters with a single parameter that has the four possible values listed in the documentation above would help.

Comment (JasonFengJ9) [reply runtime/vm/CRIUHelpers.cpp]: Replaced the two bool parameters with
```
#define SUSPEND_THEADS_NO_DELAY_HALT_FLAG 0
#define SUSPEND_THEADS_WITH_DELAY_HALT_FLAG 1
#define RESUME_THREADS_WITH_DELAY_HALT_FLAG 2
#define RESUME_THREADS_NO_DELAY_HALT_FLAG 3
```

**Comment by JasonFengJ9** [comment]
> I think the result looks good, but I suggest this should be squashed

Squashed

**Comment by keithc-ca** [comment]
Please reformat the commit message to avoid overly long lines; see https://github.com/eclipse/omr/blob/master/CONTRIBUTING.md#commit-guidelines for guidance.

**Comment by JasonFengJ9** [comment]
> Please reformat the commit message to avoid overly long lines;

Updated the commit message.

**Comment by keithc-ca** [comment]
Jenkins test sanity alinux jdk21

**Pull request merged**